discussion/ownership-and-mutable-aliases.tex

```diff
@@ -1,6 +1,103 @@
+
+\begin{enumerate}[\hspace{13}1)]
+    \item Compile-time verification can guarantee that the peripherals gets initialized correctly.
+    \item The peripherals are owned by a variable, and access to these peripherals can be controlled by {\rust}'s borrow-checker.
+\end{enumerate}
+
+Point 1) is indeed a very interesting feature of {\zinc}, but it does not really help us to solve the problem that we are discussing in this section.
+Point 2) can help us to make sure that all peripheral access goes through the \code{run\_args} structure, but this structure is only passed to the program's main loop.
+What happens if we want to use a peripheral in an interrupt handler, which defines a different scope that is outside of the main loop, like in the example from the previous section?
+As {\zinc} puts it on the project their project page; ``Accessing hardware directly is \emph{unsafe}, but you can do that as well if you want''.
+
+\subsection{Limitations with the Bindings}
+
+Our approach to define a software library for the {\gecko} has been to utilize what \emph{already} existed.
+We decided to write {\rust}-bindings for {\emlib} because it would get the project up an running fairly quickly.
+This choice also meant that we were able to quickly use a wide range of peripherals, and implement a couple of projects to test {\rust} in an embedded setting.
+It turns out, however, that this approach has introduced a few limitations to our system:
+
+\begin{enumerate}[\hspace{13}1)]
+    \item We access the peripherals through {\C}-bindings, which mean that {\rust} have no control over what happens on the other side of the \gls{ffi}.
+
+    \item Each peripheral structure that are used throughout the program is essentially a reference to a singleton \gls{mmio}.
+    This means that the peripherals can be treated as \emph{static mutable} objects, which by definition is unsafe in {\rust}.
+
+\end{enumerate}
+% It turns out that our approach to start programming th
+
+The biggest problem with both 1) and 2) is that the bindings hide away the fact that something unsafe is happening underneath the function calls, and that some of the functions expose \emph{internal mutable state}.
+An immediate proposal to fix some of the flaws in this design is:
+
+\begin{enumerate}[\hspace{13}1)]
+    \item Mark the {\rust}-bindings that instantiates and returns a peripheral as \texttt{unsafe}.
+    The other bindings do not need to be marked as unsafe because we trust them.
+
+    \item Modify the bindings that writes to (i.e. mutates) the peripheral control registers to require a \emph{mutable reference} to the respective peripheral.
+\end{enumerate}
+
+By implementing 1) across the bindings-library we will enforce the programmer to wrap every piece of code that is used to instantiate access to a peripheral with an \code{unsafe} block.
+This does not solve any problems directly, but it provokes the programmer to take extra caution when using the library.
```

Comment:
Noen tanker:

`This moves the responsibility of analyzing where mutable aliases occurs to the programmer. As this is an application dependent issue, this might be the right decision for the low level library the emlib bindings are intended to be. An auxiliary library can then be defined to handle the issue in an opinionated fashion, which might apply to alot of application, but not all.`